### ISSUE TITLE
CGALConfig.cmake does not correctly handle /lib -> /usr/lib symlink (solution: use `file(REAL_PATH)` from CMake 3.19)

### ISSUE THREAD

**Issue by scgtrp** [CGAL/cgal#8521]
TL;DR: Read https://github.com/CGAL/cgal/issues/8521#issuecomment-4031207077 and the next comment for an analysis of the issue, and a proposal for a patch. -- @lrineau, 2026/03/10

---

## Issue Details

I'm on Arch, which does not use a separate /usr partition and replaces /lib with a symlink to /usr/lib.

I'm trying to compile OpenSCAD. It finds CGAL's cmake configuration at `/lib/cmake/CGAL/CGALConfig.cmake`. Then:
- `get_filename_component(CGAL_CONFIG_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)` sets `CGAL_CONFIG_DIR=/lib/cmake/CGAL`
- the chain of `get_filename_component(CGAL_ROOT "${CGAL_ROOT}" DIRECTORY)` calls sets `CGAL_ROOT=/`
- `CGAL_FOUND` is set by testing for the existence of `/include/CGAL/config.h`, which does not exist because `/include` isn't a thing
- and OpenSCAD's build incorrectly complains that I don't have CGAL installed.

My CGAL installation otherwise seems fine; bodging it with
```cmake
get_filename_component(CGAL_CONFIG_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
set(CGAL_CONFIG_DIR "/usr${CGAL_CONFIG_DIR}")
```
is enough to allow the build to succeed.

## Environment

* Operating system: Arch Linux, amd64
  * [Arch does not appear to make any changes to the upstream cmake files.](https://gitlab.archlinux.org/archlinux/packaging/packages/cgal/-/blob/main/PKGBUILD)
* Compiler: gcc 14.2.1
* Release or debug mode:
* Specific flags used (if any):
* CGAL version: 6.0
* Boost version: 1.86.0
* Other libraries versions if used (Eigen, TBB, etc.):

**Comment by mglisse** [comment]
Any idea why cmake looks for the file in /lib/cmake and not the equivalent and more canonical /usr/lib/cmake? On a debian system with the same /lib -> /usr/lib symlink, I do get a path starting with /usr/lib. Maybe something in your environment, or a strange cmake config in arch?
I am unsure what the implication would be of adding `REALPATH` to one of the get_filename_component.

**Comment by mglisse** [comment]
When you tried again using a PATH that has /usr/bin before (or instead of) /bin, was it with a clean directory, or did it have remnants of previous tries?
I can indeed reproduce the issue if I make my PATH start with /bin. This looks like an unintended consequence of the links to /usr, I think it would make sense for cmake itself to deal with it (either upstream or in distribution-specific patches) by replacing `/` with `/usr` as a prefix candidate when /lib is a symlink to `/usr/lib`.

**Comment by lrineau** [comment]
So, eventually, @scgtrp, do you still think that is a bug from CGAL?

**Comment by lrineau** [comment]
That seems fine. I think I agree with you, but I am busy on other things.

@scgtrp @mglisse Does any of you has a patch we can commit? Maybe a pull-request?

**Comment by stinovlas** [comment]
Hi! I just wanted to point out that this is still very much a problem today. I don't really know much about `cmake`, but I'll try to summarize the problem as requested:

When `/bin` is present in `PATH` before `/usr/bin` AND `/lib` is a symlink to `/usr/bin`, cmake will use `/lib/cmake/CGAL/CGALConfig.cmake`, which fails to detect the correct correct cgal directories. This results in setting `CGAL_FOUND` to `FALSE` and failing the build.

Proposed solution: resolve symlinks in `/lib/cmake/CGAL/CGALConfig.cmake` (e.g. using `realpath`).

If you don't have a time for this, but you're willing to make the change happen, I can study a bit about `cmake` and attempt to make a PR.

**Comment by lrineau** [comment]
> `/lib` is a symlink to `/usr/bin`,

typo: `lib` is symlink to `/usr/lib`, but I agree with you anyway.

On a Linux Fedora 43 as well:

```shell
[lrineau@fernand]~% ls -1dl /*bin /lib* /usr/lib* /usr/*bin
lrwxrwxrwx. 1 root root      7 Jul 30  2025 /bin -> usr/bin
lrwxrwxrwx. 1 root root      7 Jul 30  2025 /lib -> usr/lib
lrwxrwxrwx. 1 root root      9 Jul 30  2025 /lib64 -> usr/lib64
lrwxrwxrwx. 1 root root      8 Jul 30  2025 /sbin -> usr/sbin
dr-xr-xr-x. 1 root root 122598 Mar 10 09:23 /usr/bin
dr-xr-xr-x. 1 root root  53656 Mar  9 10:08 /usr/lib
dr-xr-xr-x. 1 root root 372504 Mar 10 09:23 /usr/lib64
drwxr-xr-x. 1 root root  11658 Mar 10 09:22 /usr/libexec
lrwxrwxrwx. 1 root root      3 Apr 29  2025 /usr/sbin -> bin
```

But the bug you describe does not apply to Fedora, because `CGALConfig.cmake` is in `/usr/share/cmake/CGAL`, and not in `/usr/lib/*`. I will try to reproduce in a container.

_to be continued_

**Comment by lrineau** [comment]
Actually that is very easy to reproduce:
  1. Add a symbolic link in `/usr/lib/cmake`
  2. Force `CGAL_DIR`

```cmake
[lrineau@fernand]~/tmp/test-CGAL% ls -l /lib/cmake/CGAL                                                                   
lrwxrwxrwx. 1 root root 21 Mar 10 13:54 /lib/cmake/CGAL -> /usr/share/cmake/CGAL
[lrineau@fernand]~/tmp/test-CGAL% cat CMakeLists.txt                                                                      
cmake_minimum_required(VERSION 3.31)
project(test-CGAL)
find_package(CGAL REQUIRED)
[lrineau@fernand]~/tmp/test-CGAL% CGAL_DIR=/lib/cmake/CGAL cmake . -B build --trace-expand --trace-source=CGALConfig.cmake
Put cmake in trace mode, but with variables expanded.
Put cmake in trace mode, but output only lines of a specified file. Multiple options are allowed.
/lib/cmake/CGAL/CGALConfig.cmake(6):  set(CGAL_REQUESTED_COMPONENTS  )
/lib/cmake/CGAL/CGALConfig.cmake(8):  set(CGAL_LIBRARIES CGAL )
/lib/cmake/CGAL/CGALConfig.cmake(10):  get_filename_component(CGAL_CONFIG_DIR /lib/cmake/CGAL/CGALConfig.cmake PATH )
/lib/cmake/CGAL/CGALConfig.cmake(12):  function(cgal_detect_branch_build VAR_NAME )
/lib/cmake/CGAL/CGALConfig.cmake(20):  set(CGAL_FOUND FALSE )
/lib/cmake/CGAL/CGALConfig.cmake(21):  cgal_detect_branch_build(BRANCH_BUILD )
/lib/cmake/CGAL/CGALConfig.cmake(13):  if(IS_DIRECTORY /lib/cmake/CGAL/../../../../Installation/package_info/Installation/ )
/lib/cmake/CGAL/CGALConfig.cmake(15):  else()
/lib/cmake/CGAL/CGALConfig.cmake(16):  set(BRANCH_BUILD FALSE PARENT_SCOPE )
/lib/cmake/CGAL/CGALConfig.cmake(22):  if(BRANCH_BUILD )
/lib/cmake/CGAL/CGALConfig.cmake(43):  else()
/lib/cmake/CGAL/CGALConfig.cmake(44):  include(/lib/cmake/CGAL/CGALConfig-installation-dirs.cmake OPTIONAL RESULT_VARIABLE _has_installation_dirs )
/lib/cmake/CGAL/CGALConfig.cmake(45):  if(NOT _has_installation_dirs )
/lib/cmake/CGAL/CGALConfig.cmake(46):  set(CGAL_ROOT /lib/cmake/CGAL )
/lib/cmake/CGAL/CGALConfig.cmake(47):  get_filename_component(CGAL_ROOT /lib/cmake/CGAL DIRECTORY )
/lib/cmake/CGAL/CGALConfig.cmake(48):  if(NOT EXISTS /lib/cmake/include/CGAL/config.h )
/lib/cmake/CGAL/CGALConfig.cmake(49):  get_filename_component(CGAL_ROOT /lib/cmake DIRECTORY )
/lib/cmake/CGAL/CGALConfig.cmake(51):  if(NOT EXISTS /lib/include/CGAL/config.h )
/lib/cmake/CGAL/CGALConfig.cmake(52):  get_filename_component(CGAL_ROOT /lib DIRECTORY )
/lib/cmake/CGAL/CGALConfig.cmake(54):  if(NOT EXISTS //include/CGAL/config.h )
/lib/cmake/CGAL/CGALConfig.cmake(55):  get_filename_component(CGAL_ROOT / DIRECTORY )
/lib/cmake/CGAL/CGALConfig.cmake(59):  if(EXISTS /lib/cmake/CGAL/CGAL_add_test.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(61):  set(CGAL_MODULES_DIR /lib/cmake/CGAL )
/lib/cmake/CGAL/CGALConfig.cmake(67):  set(CGAL_INSTALLATION_PACKAGE_DIR / )
/lib/cmake/CGAL/CGALConfig.cmake(68):  set(CGAL_GRAPHICSVIEW_PACKAGE_DIR / )
/lib/cmake/CGAL/CGALConfig.cmake(69):  set(CGAL_INCLUDE_DIRS //include )
/lib/cmake/CGAL/CGALConfig.cmake(70):  if(EXISTS //include/CGAL/config.h )
/lib/cmake/CGAL/CGALConfig.cmake(76):  if(NOT CGAL_DATA_DIR )
/lib/cmake/CGAL/CGALConfig.cmake(77):  if(DEFINED ENV{CGAL_DATA_DIR} )
/lib/cmake/CGAL/CGALConfig.cmake(79):  else()
/lib/cmake/CGAL/CGALConfig.cmake(80):  if(EXISTS //Data/data )
/lib/cmake/CGAL/CGALConfig.cmake(82):  else()
/lib/cmake/CGAL/CGALConfig.cmake(83):  if(EXISTS //data )
/lib/cmake/CGAL/CGALConfig.cmake(85):  else()
/lib/cmake/CGAL/CGALConfig.cmake(86):  if(EXISTS /home/lrineau/tmp/test-CGAL/../data )
/lib/cmake/CGAL/CGALConfig.cmake(88):  else()
/lib/cmake/CGAL/CGALConfig.cmake(89):  if(EXISTS /home/lrineau/tmp/test-CGAL/../../data )
/lib/cmake/CGAL/CGALConfig.cmake(91):  else()
/lib/cmake/CGAL/CGALConfig.cmake(92):  if(RUNNING_CGAL_AUTO_TEST OR CGAL_TEST_SUITE )
/lib/cmake/CGAL/CGALConfig.cmake(102):  if(NOT TARGET CGAL::Data )
/lib/cmake/CGAL/CGALConfig.cmake(103):  add_library(CGAL::Data INTERFACE IMPORTED GLOBAL )
/lib/cmake/CGAL/CGALConfig.cmake(104):  if(NOT  STREQUAL  )
/lib/cmake/CGAL/CGALConfig.cmake(110):  include(/lib/cmake/CGAL/CGAL_CreateSingleSourceCGALProgram.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(111):  include(/lib/cmake/CGAL/CGAL_Macros.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(112):  include(/lib/cmake/CGAL/CGAL_Common.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(113):  include(/lib/cmake/CGAL/CGAL_TweakFindBoost.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(114):  include(/lib/cmake/CGAL/CGAL_enable_end_of_configuration_hook.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(116):  include(/lib/cmake/CGAL/CGALConfigVersion.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(119):  cgal_setup_module_path()
/lib/cmake/CGAL/CGALConfig.cmake(121):  include(/lib/cmake/CGAL/CGAL_target_use_TBB.cmake )
/lib/cmake/CGAL/CGALConfig.cmake(123):  if(CGAL_DEV_MODE OR RUNNING_CGAL_AUTO_TEST OR CGAL_TEST_SUITE )
/lib/cmake/CGAL/CGALConfig.cmake(130):  if(GNU MATCHES AppleClang AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 11.0.3 )
/lib/cmake/CGAL/CGALConfig.cmake(136):  foreach(comp  )
/lib/cmake/CGAL/CGALConfig.cmake(147):  set(CGALConfig_all_targets_are_defined TRUE )
/lib/cmake/CGAL/CGALConfig.cmake(148):  foreach(cgal_lib CGAL )
/lib/cmake/CGAL/CGALConfig.cmake(149):  if(TARGET CGAL::CGAL )
/lib/cmake/CGAL/CGALConfig.cmake(151):  else()
/lib/cmake/CGAL/CGALConfig.cmake(152):  set(CGALConfig_all_targets_are_defined FALSE )
/lib/cmake/CGAL/CGALConfig.cmake(156):  if(NOT CGAL_FOUND )
/lib/cmake/CGAL/CGALConfig.cmake(157):  return()
CMake Error at CMakeLists.txt:3 (find_package):
  Found package configuration file:

    /lib/cmake/CGAL/CGALConfig.cmake

  but it set CGAL_FOUND to FALSE so package "CGAL" is considered to be NOT
  FOUND.


```

The issue is that `CGAL_DIR` is `/lib/cmake/CGAL` and then the CMake code searches for CGAL headers in `/lib/cmake/../../include` (which is the non-existent `/include`).

**Comment by lrineau** [comment]
The CMake way to solve this is [`file(REAL_PATH ...)`](https://cmake.org/cmake/help/latest/command/file.html#real-path). I will write a patch.

**Comment by rafayali4u** [comment]
Hi, I am interested in working on this issue.

It seems that the problem is related to incorrect handling of symbolic links (e.g., /lib -> /usr/lib) in CGALConfig.cmake, which may lead to incorrect path resolution during build.

I plan to investigate how paths are currently resolved and propose a fix using file(REAL_PATH) or similar CMake utilities to ensure correct behavior across systems.

Please let me know if this approach aligns with your expectations or if there are specific constraints I should consider.

**Comment by lrineau** [comment]
@scgtrp @stinovlas The PR https://github.com/CGAL/cgal/pull/9390 is now merged in the `main` branch of CGAL, solving this issue. Could you check if that actually solve your problem?

**Comment by ziyao233** [comment]
> [@scgtrp](https://github.com/scgtrp) [@stinovlas](https://github.com/stinovlas) The PR [#9390](https://github.com/CGAL/cgal/pull/9390) is now merged in the `main` branch of CGAL, solving this issue. Could you check if that actually solve your problem?

I maintain another usr-merged distribution, and could confirm backporting #9390 to v6.2 resolves the issue when building OpenSCAD.

**Comment by sloriot** [comment]
The fix was merged only in the main branch, targeting 6.3

**Comment by lrineau** [comment]
> The fix was merged only in the main branch, targeting 6.3

I think I should backport it: the bug was reported against 6.1.x.

**Comment by sloriot** [comment]
Backport done in 6.1x and 6.2x
